relay/blob_manager.go

```diff
@@ -0,0 +1,78 @@
+package relay
+
+import (
+	"context"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigenda/disperser/common/v2/blobstore"
+	"github.com/Layr-Labs/eigenda/relay/cache"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+)
+
+// blobManager encapsulates logic for fetching blobs. Utilized by the relay Server.
+// This struct adds caching and threading on top of blobstore.BlobStore.
+type blobManager struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// blobStore can be used to read blobs from S3.
+	blobStore *blobstore.BlobStore
+
+	// blobCache is an LRU cache of blobs.
+	blobCache cache.CachedAccessor[v2.BlobKey, []byte]
+
+	// concurrencyLimiter is a channel that limits the number of concurrent operations.
+	concurrencyLimiter chan struct{}
+}
+
+// newBlobManager creates a new blobManager.
+func newBlobManager(
+	ctx context.Context,
+	logger logging.Logger,
+	blobStore *blobstore.BlobStore,
+	blobCacheSize int,
+	workPoolSize int) (*blobManager, error) {
+
+	server := &blobManager{
+		ctx:                ctx,
+		logger:             logger,
+		blobStore:          blobStore,
+		concurrencyLimiter: make(chan struct{}, workPoolSize),
+	}
+
+	cache, err := cache.NewCachedAccessor[v2.BlobKey, []byte](blobCacheSize, server.fetchBlob)
+	if err != nil {
+		return nil, fmt.Errorf("error creating blob cache: %w", err)
+	}
+	server.blobCache = cache
+
+	return server, nil
+}
+
+// GetBlob retrieves a blob from the blob store.
+func (s *blobManager) GetBlob(blobKey v2.BlobKey) ([]byte, error) {
+
+	s.concurrencyLimiter <- struct{}{}
+	data, err := s.blobCache.Get(blobKey)
+	<-s.concurrencyLimiter
+
+	if err != nil {
+		// It should not be possible for external users to force an error here since we won't
+		// even call this method if the blob key is invalid (so it's ok to have a noisy log here).
+		s.logger.Error("Failed to fetch blob: %v", err)
+		return nil, err
+	}
+
+	return *data, nil
+}
+
+// fetchBlob retrieves a single blob from the blob store.
+func (s *blobManager) fetchBlob(blobKey v2.BlobKey) (*[]byte, error) {
+	data, err := s.blobStore.GetBlob(s.ctx, blobKey)
+	if err != nil {
+		s.logger.Error("Failed to fetch blob: %v", err)
+		return nil, err
+	}
+
+	return &data, nil
```

Comment:
Is this pointer needed? The slice is already a pointer

Reply:
fixed